olive/passes/onnx/conversion.py

```diff
@@ -420,34 +421,41 @@ def _convert_distributed_model_on_device(
         torch_dtype: Optional[torch.dtype] = None,
     ) -> DistributedOnnxModel:
         """Convert a DistributedPyTorchModel to a DistributedOnnxModel."""
-        from mpi4py.futures import MPIPoolExecutor
-
         pass_config = config
         model_config = model.to_json()["config"]
         world_size = model.num_ranks
         output_model_path = str(Path(output_model_path).with_suffix(""))
+        use_gpu = self.accelerator_spec.accelerator_type == Device.GPU
 
         params = [
             (
                 pass_config,
                 model_config,
-                world_size,
-                device,
+                torch.device("cuda", rank) if use_gpu else torch.device("cpu"),
                 torch_dtype,
                 rank,
                 output_model_path,
+                tempfile.tempdir,
             )
             for rank in range(world_size)
         ]
 
-        with MPIPoolExecutor(max_workers=world_size) as executor:
-            executor.map(OnnxConversion._export_ranked_model, params)
-            executor.shutdown()
+        max_parallel_jobs = min(world_size, config["parallel_jobs"] or multiprocessing.cpu_count())
+        if max_parallel_jobs <= 1:
+            results = [OnnxConversion._export_ranked_model(_) for _ in params]
+        else:
+            context = multiprocessing.get_context("spawn")
+            with context.Pool(processes=max_parallel_jobs) as pool:
+                results = pool.map(OnnxConversion._export_ranked_model, params)
+
```

Comment:
In case of multiprocessing jobs, the GPU is not dedicated to any specific process. The spawned processes could in theory be on separate CPU.

In the worst-case scenario, parallel jobs could be set to zero to run synchronously.

Reply:
just to be clear, is the multiprocessing option available:
-  only to ensure we don't overwhelm the cpu by running all jobs in parallel on a slower machine?
- Also to allow sharding a large model using fewer GPUs available than needed to do them in parallel?;
  - If this is the case, should the device id of the sharded model be something like process_idx % max_parallel_jobs? 
  
From the code, looks like it is only for the first scenario. If so, should we check beforehand that `torch.cuda.device_count() <= world_size`?

Reply:
FYI for the second scenario https://stackoverflow.com/questions/10190981/get-a-unique-id-for-worker-in-python-multiprocessing-pool